commit cd1101e8c87aa175c2d5e87ddec656e50ef4ab5d
Author: Christoph M. Becker <cmbecker69@gmx.de>
Date:   Mon Jul 29 16:08:03 2019 +0200

    Fix #77919: Potential UAF in Phar RSHUTDOWN
    
    We have to properly clean up in case phar_flush() is failing.
    
    We also make the expectation of the respective test case less liberal
    to avoid missing such bugs in the future.

diff --git a/ext/phar/phar_object.c b/ext/phar/phar_object.c
index 6bd5c1dc33..c1ba97a195 100644
--- a/ext/phar/phar_object.c
+++ b/ext/phar/phar_object.c
@@ -2037,7 +2037,7 @@ static zend_object *phar_rename_archive(phar_archive_data **sphar, char *ext, ze
 	char *newname = NULL, *newpath = NULL;
 	zval ret, arg1;
 	zend_class_entry *ce;
-	char *error;
+	char *error = NULL;
 	const char *pcr_error;
 	int ext_len = ext ? strlen(ext) : 0;
 	size_t new_len, oldname_len;
@@ -2205,6 +2205,8 @@ its_ok:
 	phar_flush(phar, 0, 0, 1, &error);
 
 	if (error) {
+		zend_hash_str_del(&(PHAR_G(phar_fname_map)), newpath, phar->fname_len);
+		*sphar = NULL;
 		zend_throw_exception_ex(spl_ce_BadMethodCallException, 0, "%s", error);
 		efree(error);
 		efree(oldpath);
